### ISSUE TITLE
Filesystem Helper API's

### ISSUE THREAD

**Issue by Redth** [xamarin/Caboodle#14]
Some simple and helpful abstractions for common Filesystem problems

Possible functionality:
 - Find Paths
   - Cache
   - Sync'd/Roaming
   - Data
   - User / Document

**Comment by mattleibow** [comment]
Would this be something that would exist in a nested class under the Platform API:

```csharp
public class Platform {
    public class Directories {
        public static string AppData { get; }
    }
}
```

**Comment by Redth** [comment]
Yeah my first thought was stick it in Platform as well.  Can we think of any other interesting bits we might be able to provide related to filesystem?

**Comment by mattleibow** [comment]
I removed `Roaming` as this is not really supported the same way across devices. For example, on iOS the user must have an iCloud account and agree to use it.

This concept of synchronised data should be a separate one. This API is for local-only - even though it is backed up, it is still just for this device.

**Comment by sparkie108** [comment]
The design should be validated against UWP, for example [ApplicationData.Current.LocalCacheFolder](https://docs.microsoft.com/en-us/uwp/api/Windows.Storage.ApplicationData#Windows_Storage_ApplicationData_LocalCacheFolder) returns a StorageFolder object not a string.

**Comment by jamesmontemagno** [comment]
Correct, it would just be the .Path at the end :)

**Comment by sparkie108** [comment]
if only it was that easy.  If you try the following code:
 ```
var tempDirPath = ApplicationData.Current.LocalCacheFolder.Path;
var tempFilePath = Path.Combine(tempDirPath, "test.txt");
File.WriteAllText(tempDirPath, "Hello World!");
```
The write throws a System.UnauthorizedAccessException.  The only way I could get writing temp files to work was using the StorageFolder/StorageFile classes.

**Comment by mattleibow** [comment]
Thanks for bringing this up, I will investigate.

**Comment by sparkie108** [comment]
Just been retesting this and noticed a typo in the quick test I did yesterday (oops)
```
var tempDirPath = ApplicationData.Current.LocalCacheFolder.Path;
var tempFilePath = Path.Combine(tempDirPath, "test.txt");
File.WriteAllText(tempFilePath, "Hello World!");
```
and it now works!  Looks like they fixed it in the .NET Std 2.0 update (which is what you are targeting).
So the proposed design should work fine.
